crates/ruff_linter/src/rules/refurb/rules/hardcoded_string_charset.rs

```diff
@@ -0,0 +1,168 @@
+use ruff_macros::{derive_message_formats, violation};
+use ruff_python_ast::{CmpOp, Expr, ExprStringLiteral, StringLike};
+use ruff_text_size::{Ranged, TextRange};
+
+/// ## What it does
+/// Checks for uses of hardcoded charsets, which are defined in Python string module.
+///
+/// ## Why is this bad?
+/// Usage of named charsets from the standard library is more readable and less error-prone.
+///
+/// ## Example
+/// ```python
+/// x = "0123456789"
+/// y in "abcdefghijklmnopqrstuvwxyz"
+/// ```
+///
+/// Use instead
+/// ```python
+/// import string
+///
+/// x = string.digits
+/// y in string.ascii_lowercase
+/// ```
+///
+/// ## References
+/// - [Python documentation: String constants](https://docs.python.org/3/library/string.html#string-constants)
+#[violation]
+pub struct HardcodedStringCharset {
+    name: &'static str,
+}
+
+impl AlwaysFixableViolation for HardcodedStringCharset {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Use of hardcoded string charset")
+    }
+
+    fn fix_title(&self) -> String {
+        let HardcodedStringCharset { name } = self;
+        format!("Replace hardcoded charset with `string.{name}`")
+    }
+}
+
+struct Charset {
+    name: &'static str,
+    bytes: &'static [u8],
+    bitset: u128,
+}
+
+impl Charset {
+    const fn new(name: &'static str, value: &'static str) -> Self {
+        let bytes = value.as_bytes();
+
+        // let bitset = bytes.iter().fold(0, |acc, &byte| acc | (1 << byte));
+        // TODO: replace with ^ or for-loop, when these features will be supported in `const` fn
+        //  - https://github.com/rust-lang/rust/issues/67792
+        //  - https://github.com/rust-lang/rust/issues/87575
+        let mut bitset = 0;
+        let mut i = 0;
+        while i < bytes.len() {
+            bitset |= 1 << bytes[i];
+            i += 1;
+        }
+
+        Self {
+            name,
+            bytes,
+            bitset,
+        }
+    }
+}
+
+const KNOWN_CHARSETS: [Charset; 9] = [
+    Charset::new(
+        "ascii_letters",
+        "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ",
+    ),
+    Charset::new("ascii_lowercase", "abcdefghijklmnopqrstuvwxyz"),
+    Charset::new("ascii_uppercase", "ABCDEFGHIJKLMNOPQRSTUVWXYZ"),
+    Charset::new("digits", "0123456789"),
+    Charset::new("hexdigits", "0123456789abcdefABCDEF"),
+    Charset::new("octdigits", "01234567"),
+    Charset::new("punctuation", "!\"#$%&'()*+,-./:;<=>?@[\\]^_`{|}~"),
+    Charset::new(
+        "printable",
+        "0123456789abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ!\"\
+        #$%&'()*+,-./:;<=>?@[\\]^_`{|}~ \t\n\r\x0b\x0c",
+    ),
+    Charset::new("whitespace", " \t\n\r\x0b\x0c"),
+];
+
+fn check_charset_as_set(s: &str) -> Option<&Charset> {
+    if !s.is_ascii() {
+        return None;
+    }
+    let bitset = s
+        .as_bytes()
+        .iter()
+        .try_fold(0, |acc, &byte| (byte < 128).then(|| acc | (1_u128 << byte)))?;
```

Comment:
I suggest that we extract a `AsciiCharSet` struct that wraps the bitset to avoid repeating the bitset calculation. It also offers an opportunity to write some documentation.

Reply:
Added.